File: LifeGame.swift

// Time Complexity : (O)n^2
// Space Complexity : (O)1
// Did this code successfully run on Leetcode : yes
//Based on the direction logic find a neighbor value.
//apply rule. Modify live value to 2 and dead cell to 3 so that we can identify lives in place.
//last replace all 2 and 3 with 0 and 1


class Solution {
    
    func gameOfLife(_ board: inout [[Int]]) {
        let lives = getLiveNeighbors(board,0,0)
        for i in 0..<board.count {
            for j in 0..<board[0].count {
                let lives = getLiveNeighbors(board,i,j)
                //Apply rule
                if board[i][j] == 1 && (lives < 2 || lives > 3){
                    board[i][j] = 2
                }else if board[i][j] == 0 && lives == 3 {
                    board[i][j] = 3
                }
            }
        }
        
        for i in 0..<board.count {
            for j in 0..<board[0].count {
                board[i][j] = board[i][j]%2
            }
        }
    }
    
    func getLiveNeighbors(_ board: [[Int]], _ row: Int, _ column: Int) -> Int {
        let m = board.count
        let n = board[0].count
        let directions = [[1,-1],[1,0],[1,1],[0,-1],[0,1],[-1,-1],[-1,0],[-1,1]]
        var count = 0
        for d in directions {
            let r = row+d[0]
            let c = column+d[1]
            if r>=0 && c>=0 && r<m && c<n && ( board[r][c] == 1 || board[r][c] == 2){
                count += 1
            }
        }
        return count
    }
}

File: life_game.py
# Time Complexity : O(m*n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
from typing import List


class Solution:
    m = 0
    n = 0
    _board = [[]]

    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        self.m = len(board)
        self.n = len(board[0])
        self._board = board

        for i in range(self.m):
            for j in range(self.n):
                neighbours = self.count_neighbours(i, j)
                if self._board[i][j] == 1 and (neighbours < 2 or neighbours > 3):
                    self._board[i][j] = 2
                elif self._board[i][j] == 0 and neighbours == 3:
                    self._board[i][j] = 3

        for i in range(self.m):
            for j in range(self.n):
                if self._board[i][j] == 2:
                    self._board[i][j] = 0
                elif self._board[i][j] == 3:
                    self._board[i][j] = 1

        print(board)

    def count_neighbours(self, r: int, c: int) -> int:
        neighbours = [[0, 1], [1, 0], [1, 1], [-1, 0], [0, -1], [1, -1], [-1, 1], [-1, -1]]
        count = 0
        for neighbour in neighbours:
            row = r + neighbour[0]
            col = c + neighbour[1]
            if 0 <= row < self.m and 0 <= col < self.n and (self._board[row][col] == 1 or self._board[row][col] == 2):
                count += 1
        return count


if __name__ == '__main__':
    _board_ = [[0, 1, 0], [0, 0, 1], [1, 1, 1], [0, 0, 0]]
    Solution().gameOfLife(_board_)
